File: library/src/main/java/com/yausername/youtubedl_android/utils/StreamGobbler.java
package com.yausername.youtubedl_android.utils;

import android.util.Log;

import com.yausername.youtubedl_android.BuildConfig;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;

public class StreamGobbler extends Thread {

    private InputStream stream;
    private StringBuffer buffer;

    private static final String TAG = "StreamGobbler";

    public StreamGobbler(StringBuffer buffer, InputStream stream) {
        this.stream = stream;
        this.buffer = buffer;
        start();
    }

    public void run() {
        try {
            Reader in = new InputStreamReader(stream, "UTF-8");
            int nextChar;
            while ((nextChar = in.read()) != -1) {
                this.buffer.append((char) nextChar);
            }
        } catch (IOException e) {
            if(BuildConfig.DEBUG) Log.e(TAG, "failed to read stream", e);
        }
    }
}

File: library/src/main/java/com/yausername/youtubedl_android/data/local/streams/StreamGobbler.kt
package com.yausername.youtubedl_android.data.local.streams

import android.util.Log
import com.yausername.youtubedl_android.BuildConfig
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

internal class StreamGobbler(private val buffer: StringBuilder, private val stream: InputStream) :
    Thread() {

    init {
        start()
    }

    override fun run() {
        try {
            InputStreamReader(stream, StandardCharsets.UTF_8).use { reader ->
                var nextChar: Int
                while (reader.read().also { nextChar = it } != -1) {
                    buffer.append(nextChar.toChar())
                }
            }
        } catch (e: IOException) {
            if (BuildConfig.DEBUG) Log.e(TAG, "failed to read stream", e)
        }
    }

    companion object {
        private val TAG = StreamGobbler::class.java.simpleName
    }
}
